### ISSUE TITLE
Diagram spacing

### ISSUE THREAD

**Issue by devaigergely81** [ELTE-Soft/txtUML#83]
When a class diagram is created without any associations, no space is allocated between neighbouring boxes:
~~~
model testmodel {
	class A;
	class B;
}
~~~~
![boxes1](https://cloud.githubusercontent.com/assets/7555116/10715944/23e09e50-7b28-11e5-83db-6ff8caebb73f.png)

Adding an association suddenly implies a lot of space (2 * box size, I guess) in between:
~~~~
model testmodel {
	class A;
	class B;
	association AB {
		1 A a;
		* B b;
	}
}
~~~~
![boxes2](https://cloud.githubusercontent.com/assets/7555116/10715950/4a4d3e36-7b28-11e5-91c8-a2ce8003be32.png)

@GregoBalu, could you summarize here the spacing algorithm so that we can find a fix?

**Comment by GregoBalu** [comment]
This is because a simple if tests for whether there are links or no. If not the whole link arrange part is not invoked (which involves putting in the corridors). This will be fixed in no-time.
On the other hand the current algorithm for corridors and sizes are:
```
box.X := box.X * 
    Math.floor(WidthOfCells * CorridorRatio * 2.0)) + 
    AmountToMoveToTheCenter.X;
box.Y := box.Y * 
    Math.floor(HeightOfCells * CorridorRatio * 2.0)) + 
    AmountToMoveToTheCenter.Y;
```
**where:** 
 *WidthOfCells* is the number of grid lines horizontally, while *HeightOfCells* is the number of grid lines vertically. *CorridorRatio* is the amount that the user wants to inflate the diagram with (statement). *AmountToMoveToTHeCenter* is a vector that I have to move the box to the center of it's cell. The dimension of a cell is based on the tallest and widest boxes.
**Note:** *Width/HeightOfCells* is at least *k+2*, where k is the maximum amount of links on a single box.

**Comment by devaigergely81** [comment]
How can the user set the CorridorRatio? (I could not find it or similar in the API.)

Why is the `* 2.0` operation? (I guess that CorridorRatio defaults to `1`, therefore the corridors become double the box sizes as seen in my first comment of this issue. That is too much.)

The rule that "the width of a cell is `>= k+2`, where `k` is the max link end on a box" was invented to provide enough slots to connect link ends. However, the real box sizes can be smaller than the cell size, therefore less than `k` slots might be available. How is this handled?

**Comment by GregoBalu** [comment]
```* 2.0``` is for the corridors themselves. For every box a corridor must follow, therefore 2 cell size should be left. At least this is what I believe and my tests did not show such strange results as stated above. Anyhow, further investigation is needed.

Yes ```CorridorRatio``` is ```1.0``` by defualt.
We agreed a few weeks back that the user can alternate these by statements: ```corridorratio(1.25)``` and ```overlaparrange(few)```. I believe these are not yet implemented in the Jtxtuml part.

First I get the ```smallest box``` in the diagram (by pixel size). Then I calculate ```pixelPerGrid``` values for vertical and horizontal: ```smallestbox.PixelWidth / (k+2)```, etc. Then for every box the Width (grid) equals the ```PixelWidth / pixelPerGrid```. After this a simple max search is done to find the greatest width and height values, which will be the actual Cell size.
The only arguable point here is the definition of the so-called ```smallest box```. Right now it's defined as the smallest area (square pixels). With this a box might not have *k* connecting point on one side but as it's area is bigger on the other side there must be more than *k*.

**Comment by devaigergely81** [comment]
If the cell width is `w` and the corridor ratio is `r`, then a row of boxes in cells `(0,0)`, `(1,0)`, `(2,0)` ... should be placed at coordinates `(0,0)`, `(w+w*r,0)`, `(w+w*r+w+w*r,0)` ..., that is, the `x` coordintates are: `0`, `w*(1+r)`, `2*w*(1+r)` ..., the general formula is: `i*w*(1+r)`. (I omitted the `AmountToMoveToCenter` part, that needs to be added, of course.)

"my tests did not show such strange results as stated above": What is the way you test?

Regarding the `corridorratio`, I think you can go on with the implementation. (It is unclear if @kovacsgabor will have time for this in the near future, but he is the right person to review the future pull request.) I suggest the syntax `@SpacingRatio(val = 0.5)`, but anyone from the team is welcome to suggest better names.
Regarding the `overlaparrange`, we still need discussions, I would not implement it right now.

Instead of the `smallest box`, the `smallest width` and the `smallest height` need to be searched for (which might come from two different boxes), and `gridCellWidth` and  `gridCellHeight` are needed instead of a single `pixelPerGrid`.
However, there is another problem we have to think about: When moving smaller boxes to the centre of their cell, it is not guaranteed any more that the sides of the boxes would fall on grid lines. For example, a grid line might enter the box 1 pixel far from its corner. The diagram will be very strange if an edge ends in that slot, almost at the corner. We have to increase the width and the height of the boxes with the minimal amount ensuring that they are integral multiple of `gridCellWidth` and  `gridCellHeight` respectively.

**Comment by GregoBalu** [comment]
I test it with my own test-env and occasionally I get myself to try certain stuff out in papyrus but rarely as it's way too much work to assemble a model and a layout in jtxtuml.

Smallest width and height will be searched for. PixelPerGrid is of course separated to height and width right now, I just didn't write it down before.

Well, currently I cheat a bit with the dimensions, as they are calculated like this:
```
foreach(box : Boxes)
  box.Width := (int) Math.ceil( box.PixelWidth / pixelPerGridWidth );
  box.Height := (int) Math.ceil( boxPixelHeight / pixelPerGridHeight );
```
And then I move the boxes in their cell to the middle by grids. So it won't be the exact middle, but almost. This is basically what you suggested.

**Comment by devaigergely81** [comment]
This issue is the proof that testing the full visualization chain is necessary. The models and diagram definitions I had to write to find out the problems reported here are <20 lines altogether, I would not name it "too much work". :) Furthermore, the models you test with can go into a test project and can be reused.

In order to get closer to the end of this issue:
* Do you agree with the formula `i*w*(1+r)` instead of `2*i*w*r`?
* Centring the boxes: Can you confirm that you also adjust the real width and height of the boxes (in pixels) according to the `box.Width` and `box.Height` shown above?

**Comment by GregoBalu** [comment]
It's time. I have to get the up-to-date trunk, then launch an eclipse import all the plugins, and from there run as a new eclipse where those are installed (2 eclipses running) and in the new create a new project, new model, new layout. As java is, ... well not so smooth, 2 eclipse is not good for the health of my laptop, and takes a lot of time to start them. Exporting the txtuml plugin is even more slower and installing from an update site is even more slower than exporting, especially when I have to recompile the whole because I modified a line in my plugin. Long story short: I test my part with my test env, which worked fine until this point, but will enforce myself to test more regularly with the whole stuff together.

1) First of all changed the formula and it didn't change anything for me. Oh now I get it. Yes your formula is ok, I was just wondering why my solution worked fine but if r == 1 it's obvious that it gives the same result. At least for me. As for papyrus the gap is still pretty large, but I think this is because the cells are large. So to get it straight this is not a fault on my side. The papyrus part doesn't even give me the pixel values yet.

2) I do not alter the pixel sizes but it can be done easily, but currently the visualization (papyrus part) calculates a size for boxes from the grid sizes (I can't recall the exact formula for this, but something like I did to get pixel->grid, but backwards. I agree that this can be moved under my guard so the visualization can just get the pixel values and work with them.
